gadget-container/hooks/nri/main.go

```diff
@@ -0,0 +1,136 @@
+package main
+
+import (
+	"context"
+	"flag"
+	"fmt"
+	"os"
+	"time"
+
+	"google.golang.org/grpc"
+
+	pb "github.com/kinvolk/inspektor-gadget/pkg/gadgettracermanager/api"
+	"github.com/kinvolk/inspektor-gadget/pkg/gadgettracermanager/containerutils"
+
+	"github.com/containerd/nri/skel"
+	types "github.com/containerd/nri/types/v1"
+)
+
+// TODO: Understand why using github.com/containerd/pkg/cri/annonations
+// creates a hell dependency problem with k8s.io packages.
+const (
+	// SandboxNamespace is the name of the namespace of the sandbox (pod)
+	SandboxNamespace = "io.kubernetes.cri.sandbox-namespace"
+
+	// SandboxName is the name of the sandbox (pod)
+	SandboxName = "io.kubernetes.cri.sandbox-name"
+
+	// ContainerName is the name of the container in the pod
+	ContainerName = "io.kubernetes.cri.container-name"
+)
+
+var (
+	socketfile string
+)
+
+func init() {
+	flag.StringVar(&socketfile, "socketfile", "/run/gadgettracermanager.socket", "Socket file")
+}
+
+type igHook struct {
+}
+
+func (i *igHook) Type() string {
+	return "ighook"
+}
+
+func (i *igHook) Invoke(ctx context.Context, r *types.Request) (*types.Result, error) {
+	// Ignore sandbox containers
+	if !r.IsSandbox() && (r.State == types.Create || r.State == types.Delete) {
+		processContainer(r)
+	}
+
+	result := r.NewResult("ighook")
+	return result, nil
+}
+
+func main() {
+	ctx := context.Background()
+	if err := skel.Run(ctx, &igHook{}); err != nil {
+		fmt.Fprintf(os.Stderr, "Error executing ighook: %v", err)
+		// don't return an error as it's a debug tool and we don't want to
+		// create extra trouble if there is a failure.
+		os.Exit(0)
+	}
+}
+
+func processContainer(r *types.Request) error {
+	// Validate state
+	if r.ID == "" || (r.Pid == 0 && r.State == types.Create) {
+		return fmt.Errorf("invalid OCI state: %v %v", r.ID, r.Pid)
+	}
+
+	// Connect to the Gadget Tracer Manager
+	var client pb.GadgetTracerManagerClient
+	var ctx context.Context
+	var cancel context.CancelFunc
+	conn, err := grpc.Dial("unix://"+socketfile, grpc.WithInsecure())
+	if err != nil {
+		return err
+	}
+	defer conn.Close()
+	client = pb.NewGadgetTracerManagerClient(conn)
+	ctx, cancel = context.WithTimeout(context.Background(), 1*time.Second)
+	defer cancel()
+
+	// Handle the poststop hook first
+	if r.State == types.Delete {
+		_, err := client.RemoveContainer(ctx, &pb.ContainerDefinition{
+			ContainerId: r.ID,
+		})
+		return err
+	}
+
+	// TODO: NRI includes a cgroup path, can that be used?
```

Comment:
I think not: NRI directly uses the cgroup path from config.json, defined by runtime-spec. And runtime-spec says the path could be relative and we have no way to know what's the base directory.

Additionally, it's not clear to me if it can be cgroup-v1 or cgroup-v2.

Source:
https://github.com/opencontainers/runtime-spec/blob/1c3f411f041711bbeecf35ff7e93461ea6789220/config-linux.md#cgroups-path

I think it is better to just check by ourselves.

But I would also like to consider removing the cgroup paths from gadget-tracer-manager altogether, because it only works on cgroup-v2, so maybe using the mntns only is more reliable anyway (in a different PR).

Reply:
Created a new issue with this idea -> https://github.com/kinvolk/inspektor-gadget/issues/178.